# Fill the Rooms project: EDA Analysis

### Financials Dataset

This dataset contains information about hotel reservations. Below, we have a description of what we can find in the dataset:

| Column                           | Description                                                                                         |
|----------------------------------|-----------------------------------------------------------------------------------------------------|
| `hotel`                          | Type of hotel                                                                                       |
| `is_canceled`                    | Indicates whether the reservation was canceled (`True`) or not (`False`)                            |
| `lead_time`                      | Number of days between the booking date and the arrival date at the hotel                           |
| `arrival_date_year`              | Year of arrival at the hotel                                                                        |
| `arrival_date_month`             | Month of arrival at the hotel                                                                       |
| `arrival_date_week_number`       | Week number of arrival at the hotel                                                                 |
| `arrival_date_day_of_month`      | Day of the month of arrival at the hotel                                                            |
| `stays_in_weekend_nights`        | Number of nights the guest stayed over the weekend                                                  |
| `stays_in_week_nights`           | Number of nights the guest stayed during the week                                                   |
| `adults`                         | Number of adults accompanying the guest in the reservation                                          |
| `children`                       | Number of children accompanying the guest in the reservation                                        |
| `babies`                         | Number of babies accompanying the guest in the reservation                                          |
| `meal`                           | Type of meal included in the reservation (`BB: Bed & Breakfast`, `HB: Half Board`, `FB: Full Board`)|
| `country`                        | Country of origin of the guest                                                                      |
| `market_segment`                 | Market segment to which the reservation belongs                                                     |
| `distribution_channel`           | Distribution channel used to make the reservation                                                   |
| `is_repeated_guest`              | Indicates whether the guest is a repeated guest (`1`) or not (`0`)                                  |
| `previous_cancellations`         | Number of reservations canceled by the guest prior to this reservation                              |
| `previous_bookings_not_canceled` | Number of reservations not canceled by the guest prior to this reservation                          |
| `reserved_room_type`             | Type of room reserved                                                                               |
| `assigned_room_type`             | Type of room assigned in the reservation                                                            |
| `booking_changes`                | Number of changes made to the reservation                                                           |
| `agent`                          | Identifier of the agent involved in the reservation                                                 |
| `company`                        | Identifier of the company involved in the reservation                                               |
| `days_in_waiting_list`           | Number of days the reservation was on the waiting list                                              |
| `customer_type`                  | Type of customer who made the reservation (`Transient`, `Contract`, `Group`, `Transient-Party`)     |
| `adr`                            | Average daily rate paid for the reservation                                                         |
| `required_car_parking_spaces`    | Number of parking spaces required by the guest                                                      |
| `total_of_special_requests`      | Total number of special requests made by the guest                                                  |
| `reservation_status`             | Reservation status (`Check-Out`: Departure, `Canceled`: Canceled)                                   |
| `reservation_status_date`        | Date of the reservation status                                                                      |


In [1]:
# Data processing
# -----------------------------------------------------------------------
import pandas as pd
import numpy as np

# Visualization
# ------------------------------------------------------------------------------
import matplotlib.pyplot as plt
import seaborn as sns

# Configuration
# -----------------------------------------------------------------------
pd.set_option('display.max_columns', None) # to be able to display all columns of the DataFrames


In [2]:
# We take a first look at the first three rows of the dataframe.
df = pd.read_csv('data/finanzas-hotel-bookings.csv', index_col=0)
df.head(3)

/tmp/ipykernel_495/1781494660.py:1: DtypeWarning: Columns (1,2,5,13,14,15,16,20,21,26,30,32) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('data/finanzas-hotel-bookings.csv', index_col=0)


hotel is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel       False      342.0             2015.0               July   
1  Resort Hotel       False      737.0                NaN               July   
2  Resort Hotel       False        7.0             2015.0               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                      27.0                        1.0   
1                      27.0                        1.0   
2                      27.0                        1.0   

   stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  \
0                      0.0                   0.0     2.0       NaN     0.0   
1                      0.0                   0.0     2.0       NaN     0.0   
2                      0.0                   1.0     1.0       0.0     0.0   

  meal country market_segment distribution_channel  is_repeated_guest  \
0   BB     PRT            NaN               Direct                0.0   
1   BB     NaN            NaN               Direct                0.0   
2   BB     GBR            NaN               Direct                0.0   

   previous_cancellations  previous_bookings_not_canceled reserved_room_type  \
0                     NaN                             0.0                  C   
1                     0.0                             0.0                NaN   
2                     0.0                             0.0                  A   

  assigned_room_type  booking_changes  agent  company  days_in_waiting_list  \
0                  C              3.0    NaN      NaN                   0.0   
1                  C              4.0    NaN      NaN                   0.0   
2                  C              0.0    NaN      NaN                   0.0   

  customer_type   adr  required_car_parking_spaces  total_of_special_requests  \
0     Transient   0.0                          0.0                        0.0   
1     Transient   0.0                          0.0                        0.0   
2     Transient  75.0                          0.0                        0.0   

  reservation_status reservation_status_date    0  
0          Check-Out     2015-07-01 00:00:00  NaN  
1          Check-Out     2015-07-01 00:00:00  NaN  
2          Check-Out     2015-07-02 00:00:00  NaN

In [3]:
# And now to the end.
df.tail(3)

hotel is_canceled   lead_time  arrival_date_year arrival_date_month  \
182874   NaN         NaN  230.689826                NaN                NaN   
182875   NaN         NaN  304.888534                NaN                NaN   
182876   NaN         NaN  341.238166                NaN                NaN   

        arrival_date_week_number  arrival_date_day_of_month  \
182874                       NaN                        NaN   
182875                       NaN                        NaN   
182876                       NaN                        NaN   

        stays_in_weekend_nights  stays_in_week_nights  adults  children  \
182874                11.409496             20.461372     NaN       NaN   
182875                16.744472             15.400773     NaN       NaN   
182876                17.791486             24.442690     NaN       NaN   

        babies meal country market_segment distribution_channel  \
182874     NaN  NaN     NaN            NaN                  NaN   
182875     NaN  NaN     NaN            NaN                  NaN   
182876     NaN  NaN     NaN            NaN                  NaN   

        is_repeated_guest  previous_cancellations  \
182874                NaN                     NaN   
182875                NaN                     NaN   
182876                NaN                     NaN   

        previous_bookings_not_canceled reserved_room_type assigned_room_type  \
182874                             NaN                NaN                NaN   
182875                             NaN                NaN                NaN   
182876                             NaN                NaN                NaN   

        booking_changes  agent  company  days_in_waiting_list customer_type  \
182874              NaN    NaN      NaN                   NaN           NaN   
182875              NaN    NaN      NaN                   NaN           NaN   
182876              NaN    NaN      NaN                   NaN           NaN   

        adr  required_car_parking_spaces  total_of_special_requests  \
182874  NaN                          NaN                        NaN   
182875  NaN                          NaN                        NaN   
182876  NaN                          NaN                        NaN   

       reservation_status reservation_status_date    0  
182874                NaN                     NaN  NaN  
182875                NaN                     NaN  NaN  
182876                NaN                     NaN  NaN

In [72]:
# Function to view the complete Dataframe information.
def get_csv_info(csv):
    # Display the DataFrame info
    print("CSV Information:\n")
    print("-----------------------------------------------------------------------")
    df.info()
    print("-----------------------------------------------------------------------")
    
    # Display the column names
    print("\nColumn Names:")
    print("-----------------------------------------------------------------------")
    print(df.columns)
    print("-----------------------------------------------------------------------")
    
    return 

# Function to view duplicate and null values in the Dataframe.
def check_duplicates_and_nulls(csv):
    # Check for duplicates
    duplicate_count = df.duplicated().sum()
    print(f"Number of duplicate rows: {duplicate_count}")
    print("-----------------------------------------------------------------------")
    
    # Check for null values
    print("\nNumber of null values per column:")
    print("-----------------------------------------------------------------------")
    print(df.isna().sum())
    print("-----------------------------------------------------------------------")
    print(f"This makes a tota of {df.isna().sum().sum()} null values.")

    return

# Function to view the descriptions of the numerical and categorical columns of the Dataframe.
def describe_dataframe(csv):
    # Describe numerical columns
    print("Numerical Columns Description:\n")
    print("-----------------------------------------------------------------------")
    print(df.describe().T)
    print("-----------------------------------------------------------------------")
    
    # Describe categorical columns
    print("\nCategorical Columns Description:\n")
    print("-----------------------------------------------------------------------")
    print(df.describe(include=['object']).T)
    print("-----------------------------------------------------------------------")
    
    return 

# Function to analyze the categorical columns of the dataframe.
def analyze_categorical_columns(csv):
    # Select categorical columns
    df_cat = df.select_dtypes(include="object")
    
    # Get the names of the categorical columns
    categorical_columns = df_cat.columns
    print(f"The categorical columns in the DataFrame are:\n {categorical_columns}")
    
    # Iterate over each categorical column to display unique values and their frequencies
    for column in categorical_columns:
        print(f"\n----------- ANALYZING THE COLUMN: '{column.upper()}' -----------\n")
        print(f"Unique values: {df_cat[column].unique()}\n")
        print(f"Frequencies of unique values:\n{df_cat[column].value_counts()}\n")
        print("-----------------------------------------------------------------------")

# Function to find the rows that have ALL nulls of the dataframe.
def find_all_null_rows(csv):
    # Identify rows where all values are null
    all_null_rows = df[df.isnull().all(axis=1)]
    
    # Display the rows with all null values
    print("Rows with all null values:\n")
    print("-----------------------------------------------------------------------")
    print(all_null_rows)
    print("-----------------------------------------------------------------------")
    
    # Optionally, get the indices of these rows
    all_null_indices = all_null_rows.index
    print("\nIndices of rows with all null values:\n", all_null_indices.tolist())
    
    return 

# Function to find negative numbers in columns of the dataframe.
def find_negative_values(csv):
    # Select numerical columns
    num_cols_df = df.select_dtypes(include=["number"])
    
    # Initialize a list to store column names with negative values
    negative_columns = []
    
    # Iterate through each numerical column to check for negative values
    for col in num_cols_df.columns:
        if (num_cols_df[col] < 0).any():
            negative_columns.append(col)
    
    # Display columns with negative values
    print("Columns with negative values:", negative_columns)
    
    # Optionally, count the negative values in a specific column, e.g., 'adr'
    if 'adr' in df.columns:
        negative_count = (df['adr'] < 0).sum()
        print(f"Number of negative values in 'adr': {negative_count}")
    
    return


def find_zero_values(csv):
    # Select numerical columns
    num_cols_df = df.select_dtypes(include=["number"])
    
    # Initialize a list to store column names with zero values
    zero_columns = []
    
    # Iterate through each numerical column to check for zero values
    for col in num_cols_df.columns:
        if (num_cols_df[col] == 0).any():
            zero_columns.append(col)
    
    # Display columns with zero values
    print("Columns with zero values:", zero_columns)
    print("-----------------------------------------------------------------------")
    
    # Optionally, display the count of zero values in each column
    for col in zero_columns:
        dtype = num_cols_df[col].dtype
        zero_count = (df[col] == 0).sum()
        print(f"Number of zero values in '{col}':\n {zero_count}\n")
        print(f"The type of the value is: {dtype}")
        print("-----------------------------------------------------------------------")
    
    return

In [74]:
get_csv_info(df)

CSV Information:

-----------------------------------------------------------------------
<class 'pandas.core.frame.DataFrame'>
Index: 182877 entries, 0 to 182876
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  object 
 2   lead_time                       119490 non-null  float64
 3   arrival_date_year               64829 non-null   float64
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        101004 non-null  float64
 6   arrival_date_day_of_month       119271 non-null  float64
 7   stays_in_weekend_nights         119490 non-null  float64
 8   stays_in_week_nights            119490 non-null  float64
 9   adults                          119428 non-null  float64
 10  children                        69302 non-null   float6

In [75]:
check_duplicates_and_nulls(df)

Number of duplicate rows: 63040
-----------------------------------------------------------------------

Number of null values per column:
-----------------------------------------------------------------------
hotel                              63487
is_canceled                        63487
lead_time                          63387
arrival_date_year                 118048
arrival_date_month                 63487
arrival_date_week_number           81873
arrival_date_day_of_month          63606
stays_in_weekend_nights            63387
stays_in_week_nights               63387
adults                             63449
children                          113575
babies                             63446
meal                               63487
country                           117823
market_segment                    122943
distribution_channel               76978
is_repeated_guest                  68501
previous_cancellations            106649
previous_bookings_not_canceled     63487
reserved_r

In [76]:
describe_dataframe(df)

Numerical Columns Description:

-----------------------------------------------------------------------
                                   count         mean         std      min  \
lead_time                       119490.0   104.172628  106.975949     0.00   
arrival_date_year                64829.0  2016.156196    0.706674  2015.00   
arrival_date_week_number        101004.0    27.175785   13.613871     1.00   
arrival_date_day_of_month       119271.0    15.795977    8.780503     1.00   
stays_in_weekend_nights         119490.0     0.939461    1.082472     0.00   
stays_in_week_nights            119490.0     2.515068    1.976511     0.00   
adults                          119428.0     6.244423   14.574814     0.00   
children                         69302.0     0.125162    0.688305     0.00   
babies                          119431.0     0.019903    0.433366     0.00   
is_repeated_guest               114376.0     0.031877    0.175674     0.00   
previous_cancellations           76228

In [77]:
analyze_categorical_columns(df)

The categorical columns in the DataFrame are:
 Index(['hotel', 'is_canceled', 'arrival_date_month', 'meal', 'country',
       'market_segment', 'distribution_channel', 'reserved_room_type',
       'assigned_room_type', 'customer_type', 'reservation_status',
       'reservation_status_date', '0'],
      dtype='object')

----------- ANALYZING THE COLUMN: 'HOTEL' -----------

Unique values: ['Resort Hotel' 'City Hotel' nan]

Frequencies of unique values:
hotel
City Hotel      79330
Resort Hotel    40060
Name: count, dtype: int64

-----------------------------------------------------------------------

----------- ANALYZING THE COLUMN: 'IS_CANCELED' -----------

Unique values: [False True nan]

Frequencies of unique values:
is_canceled
False    75166
True     44224
Name: count, dtype: int64

-----------------------------------------------------------------------

----------- ANALYZING THE COLUMN: 'ARRIVAL_DATE_MONTH' -----------

Unique values: ['July' 'August' 'September' 'October' 'Novem

In [78]:
find_all_null_rows(df)

Rows with all null values:

-----------------------------------------------------------------------
       hotel is_canceled  lead_time  arrival_date_year arrival_date_month  \
119393   NaN         NaN        NaN                NaN                NaN   
119400   NaN         NaN        NaN                NaN                NaN   
119403   NaN         NaN        NaN                NaN                NaN   
119404   NaN         NaN        NaN                NaN                NaN   
119406   NaN         NaN        NaN                NaN                NaN   
...      ...         ...        ...                ...                ...   
182772   NaN         NaN        NaN                NaN                NaN   
182773   NaN         NaN        NaN                NaN                NaN   
182774   NaN         NaN        NaN                NaN                NaN   
182775   NaN         NaN        NaN                NaN                NaN   
182776   NaN         NaN        NaN                Na

In [79]:
find_negative_values(df)

Columns with negative values: ['adr']
Number of negative values in 'adr': 1


In [80]:
find_zero_values(df)

Columns with zero values: ['lead_time', 'stays_in_weekend_nights', 'stays_in_week_nights', 'adults', 'children', 'babies', 'is_repeated_guest', 'previous_cancellations', 'previous_bookings_not_canceled', 'booking_changes', 'days_in_waiting_list', 'adr', 'required_car_parking_spaces', 'total_of_special_requests']
-----------------------------------------------------------------------
Number of zero values in 'lead_time':
 6345

The type of the value is: float64
-----------------------------------------------------------------------
Number of zero values in 'stays_in_weekend_nights':
 51998

The type of the value is: float64
-----------------------------------------------------------------------
Number of zero values in 'stays_in_week_nights':
 7645

The type of the value is: float64
-----------------------------------------------------------------------
Number of zero values in 'adults':
 369

The type of the value is: float64
------------------------------------------------------------

# EDA Analysis Conclusions

- We have 182877 rows and 32 columns.
- Of which 19 are of type float (numeric) and 13 of type object (categorical).
- The column names are already standardized.
- There are 63040 duplicate values. Some may be the same client or be a duplicate by mistake of a record.
- There are 2616271 nulls and there are in all columns.

## Unique values in Categorical

Unique values in column 'hotel':
['Resort Hotel' 'City Hotel' nan]

Unique values in column 'is_canceled':
[False True nan]

Unique values in column 'arrival_date_month':
['July' 'August' 'September' 'October' 'November' 'December' '1' 'January'
 'February' '2' '3' 'March' 'March' 'April' 'May' 'June' nan]

Unique values in the 'meal' column:
['BB' 'FB' 'HB' 'SC' 'Undefined' nan]

Unique values in column 'market_segment':
[nan 'Corporate' 'Online TA' 'Direct' 'Offline TA/TO' 'Groups'
 'Complementary' 'Undefined' 'Aviation']

Unique values in column 'distribution_channel':
['Direct' 'Corporate' 'TA/TO' nan 'Undefined' 'GDS']

Unique values in the 'reserved_room_type' column:
['C' nan 'A' 'D' 'E' 'G' 'F' 'H' 'L' 'P' 'B']

Unique values in column 'assigned_room_type':
['C' 'A' 'D' 'E' 'G' 'F' 'I' 'B' 'H' 'P' 'L' 'K' nan]

Unique values in column 'customer_type':
['Transient' nan 'Contract' 'Transient-Party' 'Group']

Unique values in the 'reservation_status' column:
['Check-Out' 'Canceled' 'No-Show' nan]

## Description of Numeric columns

- `lead_time:` There is a high variability in lead times, ranging from 0 to 737 days. This suggests that some customers book well in advance while others book at the last minute. The median of 69 days suggests that at least half of the customers book more than two months in advance.
- `arrival_date_year:` The data is concentrated in the years 2015-2017. This indicates that the dataset mainly covers these three years.
- `arrival_date_week_number:` Arrivals are distributed throughout the year, but there appears to be a concentration in the middle weeks of the year (median in week 28).
- `stays_in_weekend_nights:` Most stays include 1 weekend night, with a median of 1 night, but some records indicate up to 19 nights, suggesting possible outliers or errors.
- `stays_in_week_nights:` Weekday stays tend to be of short duration, with a median of 2 nights. However, there are outliers of up to 50 nights, which could be special cases or errors.
- `Adult:`Although the median is 2 adults, the high average suggests that there are a significant number of bookings with many adults, possibly group events. The maximum value of 59 adults is unusually high and may indicate an outlier or error in the data.
- `Children:` Most bookings do not include children (median = 0), but there are a few bookings that do not include children (median = 0), but there are a few that do not include children (median = 0).
- `Babies:`Similar to the number of children, most bookings do not include babies, but there are some extreme cases.
- `is_repeated_guest:`A small percentage of guests are repeat guests (3%).
- `previous_cancellations:`Most guests have no prior cancellations (median = 0), but there are some with a high number of cancellations, up to 26.
- `previous_bookings_not_canceled:`Similar to previous cancellations, most guests have no previous bookings that have not been canceled, but some have a high number.
- `booking_changes:` Most bookings have no changes (median = 0), but some have up to 21 changes, which could indicate uncertain planning.
- `agent:` There are many different booking agents, with high variability in their identifiers. This could reflect a diversity in booking channels.
- `company:`As with agents, there is high variability in companies, with some companies making many bookings.
- `days_in_waiting_list:`Most customers spend no time on a waiting list (median = 0), but some have waited as long as 391 days, which seems extremely unusual and could be an outlier.
- `adr (Average Daily Rate):` has a wide range, from negative values (which may indicate errors or returns) to a maximum of 5400, which seems to be a significant outlier.
- `required_car_parking_spaces:` Most guests do not require parking, but some require up to 8 spaces.
- `total_of_special_requests:` Most guests make no special requests, but some make up to 5 requests.

## Description of Categorical columns

- `hotel:` There are only two types of hotels in the dataset, with “City Hotel” being the most frequent, accounting for about 66.4% of the bookings. This indicates that the majority of bookings are made at this type of hotel.
- `is_canceled:` Most reservations were not cancelled, with `False` (not cancelled) being the most common category. Approximately 62.9% of reservations were not cancelled.
- `arrival_date_month:`There are 15 unique months, suggesting that the data may include bookings for additional months or data entry errors (such as incorrect month names). August is the most popular month, which could indicate a high season in this month.
- `Meal:` The majority of clients prefer Bed & Breakfast, accounting for over 77% of bookings. Other meal options are significantly less common.
- `country:`Portugal is the most common country of origin for clients, accounting for 40.9% of bookings. The diversity in countries (163 unique) indicates an international clientele, but there may also be a significant percentage of missing or incomplete values.
- `market_segment:` The largest market segment is “Online TA”, accounting for about 47.3% of bookings. This suggests that online travel agencies are the primary booking channel for this hotel.
- `distribution_channel:`The most common distribution channel is “TA/TO”, indicating that the majority of bookings are made through travel agents or tour operators. It accounts for more than 82% of bookings.
- `reserved_room_type:` The most common room type booked is type “A”, accounting for approximately 72% of bookings. This suggests that this room type is probably the standard or most requested room type.
- `assigned_room_type:`Similar to the reserved room, the most common assigned room is also type “A”, but there appears to be more diversity in assigned room types than reserved room types.
- `customer_type:`The majority of customers are of type “Transient”, indicating that these are customers who do not have special contracts or agreements and are likely to be individual bookings for short stays.
- `reservation_status:`The majority of reservations have ended in `Check-Out`, indicating that 62.9% of reservations were completed with the stay.
- `reservation_status_date:`There are a large number of unique dates for the reservation status, suggesting an even distribution of status update dates. The most common date is October 21, 2015, but the frequency is not extremely high.
- `Anonymous Column (0):` This appears to be a column with values that have no clear name in the data provided. “`0.0' is the most frequent value, but the usefulness of this column is unclear without further context.

## Steps Cleaning

- Eliminate rows with all nulls.
- Decide how to impute nulls.
- In `arrival_date_month` we have some numbers in a string with months by their names. Adjust this so that they are all by name.
- `reservation_status_date` should change its type to date.
- Complete the rows of `reserved_room_type` with those of `assigned_room_type` and vice versa.
- The amount of nulls in `hotel`, `is_canceled`, `arrival_date_month`, `meal`, `previous_bookings_not_canceled`, `assigned_room_type`, `days_in_waiting_list`, `adr`, `required_car_parking_spaces`, `total_of_special_requests`, `reservation_status` is the same, maybe the best is to delete those rows, because they will be the same.
- In `adr` there are negative values, but it is only 1. Change to absolute.
- Change the type of the columns `adults`, `children` and `infants` to `int` instead of `float`.
- It is very rare for `adults` to have values of 0 since a minor cannot make reservations. There should always be one. Change this.
- Round float to two decimal places.